Dev yys pa3.1 (#2)
yifeifang/ysyx_en: translate/pa3.1.md

[#](#Batch-Processing-System) Batch Processing System
=================

#### There are no bugs that cannot be fixed, only systems that we do not understand

> PA3 is a watershed for the entire PA. Starting from PA3, the complexity of the system begins to gradually increase, more and more abstraction layers will be added, the system will begin to gradually improve, and the applications will become more realistic.

Reviewer: ```suggestion
PA3 is a major milestone for the entire PA series. Starting from PA3, the complexity of the system begins to gradually increase, more and more abstraction layers will be added, the system will begin to gradually improve, and the applications will become closer to real world application.
```

This means that the bug propagation chain will become more complex, and debugging will increasingly test your understanding of system behavior. If you used to get help from others and did not understand the details of the system during the debugging process, you will find that the gap between you and others is rapidly widening in PA3: others are becoming more and more proficient in mastering the entire system, while you find it more and more difficult to move forward as you go forward. For example, the handouts and codes become increasingly difficult to understand, and the bugs become more and more difficult to understand. Getting more and more mysterious.

There is only one way to remedy the situation: you really can’t be lazy anymore.

> We have implemented a von Neumann computer system in PA2, and have run typing games and FCEUX on AM. With IOE, almost all kinds of small games can be transplanted to AM to run. These small games running mode on the computer have one characteristic that they will monopolize the entire computer system: we can always play typing games on NEMU. When we don't want to play, we will exit NEMU and re-run FCEUX to play.

Reviewer: ```suggestion
We have implemented a von Neumann computer system in PA2, and have run typing games and FCEUX on AM. With IOE, almost all kinds of small games can be ported to AM. These small games have a pattern that they will consume all the computer resources and only be run one at a time: we can always play typing games on NEMU. When we are done with it, we will exit NEMU and re-run FCEUX to play.
```

> In fact, this is how early computers worked: the system administrator loaded a specific program (actually a punch card from ancient times) into the computer, and the computer would continue to execute the program until the program ended or the administrator manually terminated it. Then the administrator manually loads the next program. The programs back then were far from as cool as the Super Mario you played, and most of them were scientific calculations and physical modeling tasks (such as ballistic trajectory calculations).

Reviewer: ```suggestion
In fact, this is how early computers worked: the system administrator loaded a specific program (actually a punch card from ancient times) into the computer, and the computer would continue to execute the program until the program ended or the administrator manually terminated it. Then the administrator manually loads the next program. The programs back then were far from as cool as the Super Mario you played, and most of them were scientific calculations and physics modeling tasks (such as ballistic trajectory calculations).
```

> Later, people thought that it would be too troublesome for the administrator to manually load new programs every time. Recalling the way the von Neumann computer works, one of its characteristics is: when the computer completes executing an instruction, Then automatically execute the next instruction. Similarly, can we let the administrator prepare a set of programs in advance, and let the computer automatically execute the next program after executing one program?

Reviewer: ```suggestion
Later, people thought that it would be too troublesome for the administrator to manually load new programs every time. Recalling the way the von Neumann computer works, one of its characteristics is: when the computer completes executing an instruction, Then automatically execute the next instruction. Similarly, can we let the administrator prepare a set of programs in advance, and let the computer automatically execute the next program after finishing the previous one?
```

That is the concept of [Batch Processing System](https://en.wikipedia.org/wiki/Batch_processing), With the batch processing system, the administrator's hands can be freed. The key to the batch processing system is to have a background program. When a foreground program ends, the background program will automatically load a new foreground program to execute.

> Such a background program is actually an operating system. Yes, you heard it right, this background program that sounds like it does nothing is really an operating system! Speaking of operating systems, you may immediately think of Windows with several GB of installation packages. But in fact, the earliest operating system [GM-NAA I](https://en.wikipedia.org/wiki/GM-NAA_I/O) put into use in history was born in 1956, and one of its main tasks is to Go to "Automatically load new programs".

Reviewer: ```suggestion
Such a background program is actually an operating system. Yes, you heard it right, this background program that sounds like it does nothing is really an operating system! Speaking of operating systems, you may immediately think of Windows with several GB of installation packages. But in fact, the earliest operating system [GM-NAA I/O](https://en.wikipedia.org/wiki/GM-NAA_I/O) put into use in history was born in 1956, and one of its main tasks is to "Automatically load new programs".
```

> #### What is an operating system? (It is recommended to think about it in the second trail)

Reviewer: ```suggestion
#### What is an operating system? (It is recommended to think about it in the second round)
```

This is a big deal, and we encourage you to come back and revisit it after taking the operating system course.

### [#](#The-simplest-operating-system) The simplest operating system

Seeing is believing, let’s first show how simple the operating system can be. The operating system used in PA is called Nanos-lite, which is a tailored version of Nanjing University’s operating system Nanos. It is an operating system tailor-made for PA . By writing Nanos-lite code, you will realize how the operating system uses the mechanisms provided by the hardware (that is, ISA and AM) to support the running of the program, which is also in line with the ultimate goal of PA. As for the full version of Nanos , you will see its true colors in the operating system course next semester.

We have prepared the framework code of Nanos-lite for everyone, which can be obtained by executing the following command:

    cd ics2023
    bash init.sh nanos-lite
    

> Nanos-lite contains all modules used in subsequent PAs, but the specific functions of most modules have not been implemented. Since the functions of the hardware (NEMU) are gradually added, Nanos-lite must also cooperate with this process. You will use some macros in `nanos-lite/include/common.h` which are used to control the functions of Nanos-lite during the experiment. As the experimental progress progresses, we will gradually explain all the modules, and Nanos-lite will do more and more work. Therefore, when reading the code of Nanos-lite, you only need to care about the modules related to the current progress, and do not get entangled in the code that is not related to the current progress.

Reviewer: ```suggestion
Nanos-lite contains all modules used in subsequent PAs, but the functionality of most modules has not yet been implemented. Since the functionality of the hardware (NEMU) are gradually added, Nanos-lite must align with it. You will use some macros in `nanos-lite/include/common.h` which controls the feature set based on your progress. As the lab progresses, we will gradually explain all the modules, and Nanos-lite will do more and more work. Therefore, when reading the code of Nanos-lite, you only need to care about the modules related to the current progress, and do not get trapped in the code that is not related to the current progress.
```

    nanos-lite
    ├── include
    │   ├── common.h
    │   ├── debug.h
    │   ├── fs.h
    │   ├── memory.h
    │   └── proc.h
    ├── Makefile
    ├── README.md
    ├── resources
    │   └── logo.txt    # Project-N logo text
    └── src
        ├── device.c    # Device abstraction
        ├── fs.c        # File System
        ├── irq.c       # Interrupt exception handling
        ├── loader.c    # Loader
        ├── main.c
        ├── mm.c        # Storage management
        ├── proc.c      # Process scheduling
        ├── ramdisk.c   # ramdisk driver
        ├── resources.S # ramdisk content and logo of Project-N
        └── syscall.c   # System call handling
    

> It should be reminded that Nanos-lite runs on AM, and AM’s APIs are available in Nanos-lite. Although the operating system is a special concept to us, from AM's perspective, it is just an ordinary C program that calls the AM API, no different from Super Mario. At the same time, you will once again realize the benefits of AM: the implementation of Nanos-lite can be architecture-independent, which means that no matter which ISA you have chosen before, you can easily run Nanos-lite, and even you can Just like developing klib, debug your Nanos-lite on `native`.

Reviewer: ```suggestion
It should be reminded that Nanos-lite runs on AM, and AM’s APIs are available in Nanos-lite. Although the operating system is a special concept to us, from AM's perspective, it is just an ordinary C program that calls the AM API, no different from Super Mario. At the same time, you will once again realize the benefits of AM: the implementation of Nanos-lite can be architecture-independent, which means that no matter which ISA you have chosen before, you can easily run Nanos-lite on it, further more you can debug your `Nanos-lite` in `native` mode just like what you did for `klib`.
```

> In addition, although it will not cause obvious misunderstandings, after the introduction of Nanos-lite, we will still use the concept of "user process" instead of "user program" in some places. If you can't understand what a process is now, you only need to understand the process as a "running program". Still can’t feel the difference between the two? Let’s give a simple example. If you open Notepad three times, there will be three Notepad processes running on the computer, but there is only one Notepad program on the disk. Process is an important concept in the operating system. Detailed knowledge about processes will be introduced in the operating system class.

Reviewer: ```suggestion
In addition, although it will not cause obvious misunderstandings, after the introduction of Nanos-lite, we will still use the concept of "guest process" instead of "guest program" in some places. If you can't understand what a process is now, you only need to understand the process as a "running program". Still can’t feel the difference between the two? Let’s give a simple example. If you open Notepad three times, there will be three Notepad processes running on the computer, but there is only one Notepad program on the disk. Process is an important concept in the operating system. Detailed knowledge about processes will be introduced in the operating system class.
```

> At the beginning, all macros related to the experimental progress in `nanos-lite/include/common.h` are not defined. At this time, the function of Nanos-lite is very simple. Let's briefly sort out the current behavior of Nanos-lite:

Reviewer: ```suggestion
At the beginning, all macros related to the experimental progress in `nanos-lite/include/common.h` are not defined. Currently, the functionality of Nanos-lite is amazingly simple. Let's briefly sort out the current behavior of Nanos-lite:
```

> 1.  Print the logo of Project-N, and output hello information and compilation time through `Log()`. It should be noted that the `Log()` macro defined in Nanos-lite is not the `Log()` macro defined in NEMU . Nanos-lite and NEMU are two independent projects, and their codes will not affect each other. You need to pay attention to this when reading the code. In Nanos-lite, the `Log()` macro outputs through the `printf()` you wrote in `klib`, which will eventually call TRM's `patch()`.

Reviewer: ```suggestion
1.  Print the logo of Project-N, and output hello information and compilation time through `Log()`. It should be noted that the `Log()` macro defined in Nanos-lite is not the `Log()` macro defined in NEMU . Nanos-lite and NEMU are two independent projects, and their codes will not affect each other. You need to pay attention to this when reading the code. In Nanos-lite, the `Log()` macro write to console through the `printf()` you wrote in `klib`, which will eventually call TRM's `patch()`.
```

2.  Call `init_device()` to perform some initialization operations on the device. Currently `init_device()` will directly call `ioe_init()`.
3.  Initialize the ramdisk. Generally speaking, the program should be stored in a permanent storage medium (such as a disk). However, simulating a disk in NEMU is a slightly complicated task, so first let Nanos-lite use a section of memory as a disk. Such a disk has a special name, called ramdisk.
4.  `init_fs()` and `init_proc()` are used to initialize the file system and create processes respectively. Currently, they do not perform meaningful operations, so you can ignore them.
5.  Call `panic()` to end the running of Nanos-lite.

Since Nanos-lite is essentially an AM program, we can compile/run Nanos-lite in the same way. Execute 

    make ARCH=$ISA-nemu run
    

in the `nanos-lite/` directory. In addition, as mentioned above, you can also compile Nanos-lite to `native` and run it to help you debug.

#### The operating system is a C program

> You may not believe it, but the .c and .h files of the framework code undoubtedly contain this ironclad fact, and there is nothing special about even the compilation method. The same is true for GNU/Linux: If you read it If you look at the source code, you will find that GNU/Linux is just a huge C program.

Reviewer: ```suggestion
You may not believe it, but the .c and .h files of the framework code undoubtedly contain this ironclad fact, and there is nothing special about even the compilation method. The same statement holds true for GNU/Linux if you look at the source code, you will find that GNU/Linux is just a huge C program.
```
